### PR TITLE
HIVE-20917: Support applyQuotesToAll option in OpenCSVSerde.

### PR DESCRIPTION
**Pull request by gigem**
This is a revised version of https://github.com/apache/hive/pull/3718 and includes only the applyQuotesToAll changes.  The other, OpenCSV changes were committed separately.

OpenCSV version 3.10 added an option called "applyQuotesToAll."  When set to false, output columns were not quoted unless they needed it. This commit adds support for that option in OpenCSVSerde.  This commit is based on a patch from Sungwoo Park <glapark@gmail.com>, which was originally based on a patch from David Engel <david@istwok.net>.

### CODE REVIEW

**Comment by deniskuzZ** [comment]
hi @gigem, in that case, could we close the  https://github.com/apache/hive/pull/3718?

File: serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java [hunk 1]
```diff
@@ -92,6 +95,8 @@ public void initialize(Configuration configuration, Properties tableProperties,
     separatorChar = getProperty(properties, SEPARATORCHAR, CSVWriter.DEFAULT_SEPARATOR);
     quoteChar = getProperty(properties, QUOTECHAR, CSVWriter.DEFAULT_QUOTE_CHARACTER);
     escapeChar = getProperty(properties, ESCAPECHAR, CSVWriter.DEFAULT_ESCAPE_CHARACTER);
+    final String temp = properties.getProperty(APPLYQUOTESTOALL);
```
Comment (deniskuzZ) [inline serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java:98]:
````
applyQuotesToAll = Boolean.parseBoolean(properties.getProperty(APPLYQUOTESTOALL))
````

**Review by deniskuzZ — COMMENTED** [review]
LGTM, some minor things

**Review by deniskuzZ — APPROVED** [review]
LGTM +1, pending tests

File: serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java [hunk 2]
```diff
@@ -92,6 +95,7 @@ public void initialize(Configuration configuration, Properties tableProperties,
     separatorChar = getProperty(properties, SEPARATORCHAR, CSVWriter.DEFAULT_SEPARATOR);
     quoteChar = getProperty(properties, QUOTECHAR, CSVWriter.DEFAULT_QUOTE_CHARACTER);
     escapeChar = getProperty(properties, ESCAPECHAR, CSVWriter.DEFAULT_ESCAPE_CHARACTER);
+    applyQuotesToAll = Boolean.parseBoolean(properties.getProperty(APPLYQUOTESTOALL))
```
Comment (deniskuzZ) [inline serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java:98]: @gigem,  ';' expected

Comment (gigem) [reply serde/src/java/org/apache/hadoop/hive/serde2/OpenCSVSerde.java]: Sorry about that.  I'm too old and should know bettter than to at least do a test compile on any changes.  It's fixed now.

**Comment by deniskuzZ** [comment]
@gigem, crtseltbl_serdeprops.q was affected by this change: 
https://ci.hive.apache.org/job/hive-precommit/job/PR-5513/4/testReport/org.apache.hadoop.hive.cli.split5/TestMiniLlapLocalCliDriver/Testing___split_21___PostProcess___testCliDriver_crtseltbl_serdeprops_/

**Comment by gigem** [comment]
> @gigem, `crtseltbl_serdeprops.q` was affected by this change: https://ci.hive.apache.org/job/hive-precommit/job/PR-5513/4/testReport/org.apache.hadoop.hive.cli.split5/TestMiniLlapLocalCliDriver/Testing___split_21___PostProcess___testCliDriver_crtseltbl_serdeprops_/

This link returns a 404 for me.  Do you have another one?  In the meantime, I'll see if I can recreate the error locally.

**Comment by deniskuzZ** [comment]
> > @gigem, `crtseltbl_serdeprops.q` was affected by this change: https://ci.hive.apache.org/job/hive-precommit/job/PR-5513/4/testReport/org.apache.hadoop.hive.cli.split5/TestMiniLlapLocalCliDriver/Testing___split_21___PostProcess___testCliDriver_crtseltbl_serdeprops_/
> 
> This link returns a 404 for me. Do you have another one? In the meantime, I'll see if I can recreate the error locally.
````
Client Execution succeeded but contained differences (error code = 1) after executing crtseltbl_serdeprops.q 
29c29
< 100|val_100
---
> "100"|"val_100"
````

**Comment by gigem** [comment]
> Client Execution succeeded but contained differences (error code = 1) after executing crtseltbl_serdeprops.q 
> 29c29
> < 100|val_100
> ---
> > "100"|"val_100"

It would seem Boolean.parseBoolean() is not parsing the empty/default property correctly.  Unfortunately, I'm not able to reproduce the error (and test any fix) because earlier tests fail in my build.  I'm a near total, Maven neophyte so could someone please tell me the magic encanation to run just the OpenCSVSerde (or Serde) tests?

**Comment by deniskuzZ** [comment]
> > ## Client Execution succeeded but contained differences (error code = 1) after executing crtseltbl_serdeprops.q
> > 29c29
> > < 100|val_100
> > > "100"|"val_100"
> 
> It would seem Boolean.parseBoolean() is not parsing the empty/default property correctly. Unfortunately, I'm not able to reproduce the error (and test any fix) because earlier tests fail in my build. I'm a near total, Maven neophyte so could someone please tell me the magic encanation to run just the OpenCSVSerde (or Serde) tests?

WDYM, there is no issue with `parseBoolean`. Aren't you changing the defaults to quote the values?
````
  public static boolean parseBoolean(String s) {
        return ((s != null) && s.equalsIgnoreCase("true"));
    }
````
to run the test:
````
mvn test -Dtest=TestMiniLlapLocalCliDriver -Dqfile=crtseltbl_serdeprops.q -Drat.skip=true
````

**Comment by gigem** [comment]
> before change
> 
> ```
> "100"|"val_100"
> ```
> 
> after change quotes are removed
> 
> ```
> 100|val_100
> ```

Ah, I see you mean now.  It was not my intent to change any original defaults.  However, this change was made a long time ago for our private use.  It was only ressurrected somewhat recently for our use with Hive/MR3 and I assumed that when working with them we'd kept any old behavior unchanged.  I'll fix this as soon as I'm able to test it properly.  I have more pressing, immediate needs to attend to first.

> @gigem, shouldn't we apply the quotes by default if `quoteChar` is provided??

I don't think so.  The desire to quote or not quote all columns can be independtent of the actual, quote character.

**Comment by deniskuzZ** [comment]
> > before change
> > ```
> > "100"|"val_100"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > after change quotes are removed
> > ```
> > 100|val_100
> > ```
> 
> Ah, I see you mean now. It was not my intent to change any original defaults. However, this change was made a long time ago for our private use. It was only ressurrected somewhat recently for our use with Hive/MR3 and I assumed that when working with them we'd kept any old behavior unchanged. I'll fix this as soon as I'm able to test it properly. I have more pressing, immediate needs to attend to first.
> 
> > @gigem, shouldn't we apply the quotes by default if `quoteChar` is provided??
> 
> I don't think so. The desire to quote or not quote all columns can be independtent of the actual, quote character.

in that case, you could simply add `applyQuotesToAll`='true' in `separator_test` table serde properties

**Comment by gigem** [comment]
> in that case, you could simply add `applyQuotesToAll`='true' in `separator_test` table serde properties

I think that would change the default behavior for unsuspecting users.

After a little bit of research, I believe the following should fix both the test and default behavior.

```
applyQuotesToAll = Boolean.parseBoolean(properties.getProperty(APPLYQUOTESTOALL, "true"));
```

My attempts to test it keep failing, though, due to unrelated build issues and I'm reluctant to commit it until I'm sure it's finally correct.

**Comment by deniskuzZ** [comment]
> > in that case, you could simply add `applyQuotesToAll`='true' in `separator_test` table serde properties
> 
> I think that would change the default behavior for unsuspecting users.
> 
> After a little bit of research, I believe the following should fix both the test and default behavior.
> 
> ```
> applyQuotesToAll = Boolean.parseBoolean(properties.getProperty(APPLYQUOTESTOALL, "true"));
> ```
> 
> My attempts to test it keep failing, though, due to unrelated build issues and I'm reluctant to commit it until I'm sure it's finally correct.

the above change makes sense to me.

**Pull request merged**